File: backend/models/system.py
from __future__ import annotations

from pydantic import BaseModel


class CpuSnapshot(BaseModel):
    model_: str  # aliased from "model" to avoid Pydantic conflict
    physical_cores: int
    logical_cores: int
    frequency_mhz: float
    frequency_min_mhz: float
    frequency_max_mhz: float
    usage_per_core: list[float]
    total_usage: float

    class Config:
        populate_by_name = True


class MemorySnapshot(BaseModel):
    total_bytes: int
    available_bytes: int
    used_bytes: int
    free_bytes: int
    percent: float
    cached_bytes: int = 0
    buffers_bytes: int = 0
    shared_bytes: int = 0
    # macOS-specific breakdown (mirrors Activity Monitor)
    wired_bytes: int = 0
    compressed_bytes: int = 0
    app_memory_bytes: int = 0
    inactive_bytes: int = 0
    purgeable_bytes: int = 0
    swap_total_bytes: int
    swap_used_bytes: int
    swap_free_bytes: int
    swap_percent: float


class SystemSnapshot(BaseModel):
    hostname: str
    os: str
    uptime_seconds: int
    cpu: CpuSnapshot
    memory: MemorySnapshot
    load_average: list[float]

File: frontend/src/types/system.ts
export interface CpuInfo {
  model: string;
  physical_cores: number;
  logical_cores: number;
  frequency_mhz: number;
  frequency_min_mhz: number;
  frequency_max_mhz: number;
  usage_per_core: number[];
  total_usage: number;
}

export interface MemoryInfo {
  total_bytes: number;
  available_bytes: number;
  used_bytes: number;
  free_bytes: number;
  percent: number;
  cached_bytes: number;
  buffers_bytes: number;
  shared_bytes: number;
  // macOS-specific breakdown (mirrors Activity Monitor)
  wired_bytes: number;
  compressed_bytes: number;
  app_memory_bytes: number;
  inactive_bytes: number;
  purgeable_bytes: number;
  swap_total_bytes: number;
  swap_used_bytes: number;
  swap_free_bytes: number;
  swap_percent: number;
}

export interface SystemInfo {
  hostname: string;
  os: string;
  uptime_seconds: number;
  cpu: CpuInfo;
  memory: MemoryInfo;
  load_average: number[];
}
